### PR TITLE
Feat/make transformer decoder callable without causal mask

### PR DESCRIPTION
**Pull request by ferraric**
Fixes https://github.com/keras-team/keras-nlp/issues/1062

### CODE REVIEW

File: keras_nlp/layers/transformer_decoder.py [hunk 1]
```diff
@@ -252,6 +252,8 @@ def call(
                 at which to update the `cross_attention_cache`. Usually, this is
                 either `0` (compute the entire `cross_attention_cache`), or
                 `None` (reuse a previously computed `cross_attention_cache`).
+             use_causal_mask: bool, defaults to True. If true, causal mask
+                (masking out future input) is applied on the decoder sequence.
```
Comment (ferraric) [inline keras_nlp/layers/transformer_decoder.py:256]: naming and documentation from previous version where this was in: https://github.com/keras-team/keras-nlp/blob/cb0fa028971475879911ddf042a1473037775ee6/keras_nlp/layers/transformer_decoder.py#L191-L192

File: keras_nlp/layers/transformer_decoder.py [hunk 2]
```diff
@@ -218,6 +217,7 @@ def call(
         self_attention_cache_update_index=None,
         cross_attention_cache=None,
         cross_attention_cache_update_index=None,
+        use_causal_mask=True,
```
Comment (ferraric) [inline keras_nlp/layers/transformer_decoder.py:220]: default to True in order to not break existing behavior

**Review by ferraric — COMMENTED** [review]
Did not add a test specifically for that since the current testing strategy does not test any of the other "_mask" arguments to the call() function. Since this argument is at the same abstraction level I decided to follow the convention

**Review by jbischof — APPROVED** [review]
Looks great thanks for the contribution!

File: keras_nlp/layers/transformer_decoder.py [hunk 3]
```diff
@@ -37,7 +37,6 @@ class TransformerDecoder(keras.layers.Layer):
     paper [Attention is All You Need](https://arxiv.org/abs/1706.03762). Users
     can instantiate multiple instances of this class to stack up a decoder.
 
-    This layer will always apply a causal mask to the decoder attention layer.
```
Comment (jbischof) [inline keras_nlp/layers/transformer_decoder.py:40]: Should we keep but change to 
```python
"""
By default this layer will apply a casual mask...
"""
```

Comment (ferraric) [reply keras_nlp/layers/transformer_decoder.py:40]: done

File: keras_nlp/layers/transformer_decoder.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:221]: sgtm. I think default to `True` is probably makes sense here regardless. The most beaten path for these decoder layers will be with the causal mask applied.

File: keras_nlp/layers/transformer_decoder.py [hunk 4]
```diff
@@ -252,6 +253,8 @@ def call(
                 at which to update the `cross_attention_cache`. Usually, this is
                 either `0` (compute the entire `cross_attention_cache`), or
                 `None` (reuse a previously computed `cross_attention_cache`).
+             use_causal_mask: bool, defaults to True. If true, causal mask
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:256]: backticks for python constants, add the missing "a" before causal mask...

```
use_causal_mask: bool, defaults to `True`. If `True`, a causal mask
```

File: keras_nlp/layers/transformer_decoder.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:257]: this lgtm, and matches the multi-head attention layer in core Keras

File: keras_nlp/layers/transformer_decoder.py [hunk 5]
```diff
@@ -306,26 +309,33 @@ def call(
...
-            if self_attention_cache_update_index is None
-            else self_attention_cache_update_index,
-        )
         decoder_mask = merge_padding_and_attention_mask(
             decoder_sequence, decoder_padding_mask, decoder_attention_mask
         )
-        if decoder_mask is not None:
-            self_attention_mask = tf.minimum(decoder_mask, self_attention_mask)
+        if use_causal_mask:
+            batch_size = tf.shape(decoder_sequence)[0]
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:316]: I think this branching is getting complex enough that we should split a private method for this

```
self_attention_mask = self._compute_self_attention_mask(
    decoder_sequence,
    self_attention_cache,
    self_attention_cache_update_index,
    use_causal_mask,
)
```

File: keras_nlp/layers/transformer_decoder.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ class TransformerDecoder(keras.layers.Layer):
     paper [Attention is All You Need](https://arxiv.org/abs/1706.03762). Users
     can instantiate multiple instances of this class to stack up a decoder.
 
-    This layer will always apply a causal mask to the decoder attention layer.
+    By default, this layer will apply a causal mask to the decoder attention layer.
```
Comment (mattdangerw) [inline keras_nlp/layers/transformer_decoder.py:40]: maybe add "This can be controlled by the `use_causal_mask` call argument.", so people know what to look for in our docs?

Comment (ferraric) [reply keras_nlp/layers/transformer_decoder.py:40]: `By default, this layer will apply a causal mask to the decoder attention layer.` from this I think `use_causal_mask` is already perfectly obvious and emphasizing it would just add unnecessary clutter to the documentation. 

Your call though, I am happy to add it if you feel like it would be helpful

File: keras_nlp/layers/transformer_decoder.py (hunk 4 above)
Comment (ferraric) [reply keras_nlp/layers/transformer_decoder.py:256]: thanks, done

File: keras_nlp/layers/transformer_decoder.py (hunk 5 above)
Comment (ferraric) [reply keras_nlp/layers/transformer_decoder.py:316]: good point, done

File: keras_nlp/layers/transformer_decoder.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:40]: no this sounds good. let's keep what we got!
